readme.md
# Health Check
## Service description
Health-Check is a simple service which allows you to monitor several web services and provide an aggregated status.  

### How it works
Each time the service is contacted it will ping (http GET) each service endpoints provided in the configuration and build an aggregated view. 
If all services respond with a status 200 the health check service will return 200 as well. 
If (at least) one of the monitored services responds with a code different than 2XX or does not respond within the expected delay (timeout) it will return 503.   
The timeout can be provided globaly and for each service.

In addition, the service is displaying the overall status of YourLoops in first position. It includes the global status (either 200 or 503 depending on the other services) and potentially the version number is passed as en environment variable `YLP_VERSION`.

The service exposes this aggregated status on route `/status`  
This end point is typically what would be provided to a load balancer to check the health of the entire stack.


## Install
Requirements:

- [Node.js](http://nodejs.org/ 'Node.js') version 10.x
- [npm](https://www.npmjs.com/ 'npm') version 6.x or higher

### Install from sources
Clone this repo [from GitHub](https://github.com/mdblp/health-check 'GitHub: health-check'), then install the dependencies:

After cloning this repository to your local machine, first make sure that you have node `10.x` and npm `6.x` installed. If you have a different major version of node installed, consider using [nvm](https://github.com/creationix/nvm 'GitHub: Node Version Manager') to manage and switch between multiple node (& npm) installations. 

Once your environment is setup with node `10.x` and npm `6.x` install the dependencies with npm:

```bash
$ npm install
```

## Running locally
### Configuration
To start properly the service expect a few environment variables to be set:
* YLP_VERSION: the version of YourLoops the health-check is running on
* MONITORED_URLS: the list of urls to ping to monitor services. These urls must be provided as a JSON array.  
  Exemple: 
  ```
    export MONITORED_URLS="[
        {\"name\": \"blip\", \"url\":\"http://localhost:3000\"},
        {\"name\": \"gatekeeper\", \"url\":\"http://localhost:9123/status\"},
        {\"name\": \"hakken\", \"url\":\"http://localhost:8000/status\"},
        {\"name\": \"highwater\", \"url\":\"http://localhost:9191/status\"}
    ]"
  ```
  You can add an optionnal timeout to each of the services (in case you need different timeout set for your monitored services):  
  ```
    export MONITORED_URLS="[
        {\"name\": \"blip\", \"url\":\"http://localhost:3000\"},
        {\"name\": \"gatekeeper\", \"url\":\"http://localhost:9123/status\",  \"pingTimeout\": 3000},
        {\"name\": \"hakken\", \"url\":\"http://localhost:8000/status\"},
        {\"name\": \"highwater\", \"url\":\"http://localhost:9191/status\,  \"pingTimeout\": 4000"}
    ]"
  ```
* SERVICE_PORT: the port on which the service should listen. Defaults to 8080.
* SERVICE_NAME: the name of the service, used in the logs. Default to "health-check".  
* PING_TIMEOUT: Global timeout value, in milliseconds. Defaults to 5000.

### Execution
```bash
$ npm start
```

You can test the service locally: 
```
$ curl http://localhost:8080/status
```
The service should return either an http status code 200, in case of success, or 503 when (at least) one of the services does not respond.  

## Contribute
Start by clonning the repo and install the dependencies as explained above. Then you are ready to change the code and test!  

### Running the tests
In order to run the unit tests, you need to have `Mocha` JS test framework installed in your environment. Run the following:
```bash
$ npm install -g mocha mocha-junit-reporter
```
For more information on Mocha, refer to their [website](https://mochajs.org/).

To run the unit tests, use:

```bash
$ npm test
```



CHANGELOG.md
# Yourloops Health Check aggregator service

Health-Check is a simple service which allows you to monitor several web services and provide an aggregated status.
The service exposes this aggregated status on route /status
This end point is typically what would be provided to a load balancer to check the health of the entire stack.

## Unreleased
### Engineering
- Modify pipeline so the publish to Operations docker registry is effective
- Modify CI parameters to accomodate changes in Jenkins library for publishing SOUP/OPENAPI

## 1.2.3 - 2020-09-30
### Engineering
- PT-1532 Base health-check image on node:10-alpine

## 1.2.2 - 2020-08-19
### Engineering
- PT-1472 Update health-check to not write logs to file

## 1.2.1 - 2020-08-04
### Engineering
- PT-1453 Generate SOUP document

## 1.2.0 - 2020-01-06
### Added
- PT-643 Augment health-check so it has its own version displayed

## 1.1.1 - 2019-11-26
### Engineering
- PT-819 Apply security patches (upgrade npm packages)

## 1.1.0 - 2019-07-31

### Added
- PT-515 Add capacity to display YourLoops version if passed as an environment variable YLP_VERSION

## 1.0.1 - 2019-07-02
### Changed
- Add a timeout to the heath check requests so a service is reported Not OK if it takes more than x seconds to respond.

## 1.0.0 - 2019-06-13
### Added
- Change the output of the service: it now returns the full list of services even when one is down.
- Return more details comming from the services themselves

## 0.1.6 - 2019-01-31
### Added
- Package the application and make Jenkins publish the package on S3.
- Create production startup scripts
- Create a version file to track the version of NodeJs to use in production
### Fixes
- Correct the npm pre-test instructions: using --only-dev did curiously prevent the installation of some development dependancies such as minimatch.

## 0.1.1 - 2018-12-18
### Features
- Initiate a new project for yourloops.  
    For now keep it simple: contact each webservices (defined by configuration), get the result and report an aggregated status and view.


service.js
/**
 * Service.js
 * Implementation of the health-check service
 */
'use strict';

var restify = require('restify');
var request = require('request-promise');
const bunyan = require("bunyan");

const responseCode = {
  success: 200,
  failure: 503
}

const responseStatus = {
  success: "OK",
  failure: "NOK"
};

class HealthCheckService {
  constructor(services, ylpVersion, hcName, hcVersion) {
    //let's check the list of services is OK:
    if (services && Array.isArray(services)) {
      services.forEach((service) => {
        if (typeof service.name != 'string' || typeof service.url != 'string') {
          throw "the list of monitored services is malformed";
        }
      });
    }
    this.services = services;
    this.ylpVersion = ylpVersion;
    this.hcName = hcName;
    this.hcVersion = hcVersion;
    //default logger
    this.logger = bunyan.createLogger({
      name: 'health-check-default',
      streams: [
        {
          level: 'info',
          stream: process.stdout
        }
      ]
    });
  }

  /**
   * check the status of a given service
   * @param {*} service the service to check
   */
  serviceStatus(service) {
    this.logger.debug("check url " + service.url);

    return new Promise((resolve, reject) => {
      var retVal = { 'service': service.name };
      // prepare request:
      var options = {
        method: 'GET',
        uri: service.url,
        timeout: service.pingTimeout,
        transform: function (body, response, resolveWithFullResponse) {
          return { 'statusCode': response.statusCode, 'data': body };
        }
      }

      request(options)
        .then((res) => {
          if (res.statusCode == 200) {
            retVal['status'] = responseStatus.success;
          } else {
            retVal['status'] = responseStatus.failure;
          }
          let details = "";
          try {
            details = JSON.parse(res.data);
          } catch (err) {
            details = res.data;
          }
          retVal['details'] = details;
          resolve(retVal);
        })
        .catch((err) => {
          retVal['status'] = responseStatus.failure;
          retVal['error'] = err.message;
          reject(retVal);
        });
    });
  }


  /**
   * Check all services (from configuration) and return the result (ok or not)
   * @param {*} req express request
   * @param {*} res express response
   */
  checkServices(req, res) {
    this.logger.info("Checking services health");
    let globalResult = [];
    let ylp = { 'service': 'yourloops', 'version': this.ylpVersion };

    // By default, we consider everything is going fine
    let resCode = responseCode.success;

    let isOK = true;
    let nbOfServices = this.services.length;
    let totalOfCheckedServices = 0;
    for (let i = 0; i < nbOfServices; i++) {
      this.serviceStatus(this.services[i]).then((result) => {
        globalResult.push(result);
      }).catch((err) => {
        isOK = false;
        globalResult.push(err);
      }).finally(() => {
        totalOfCheckedServices++;
        if (totalOfCheckedServices == nbOfServices) {
          if (isOK) {
            this.logger.info("all services are up and running!");
          } else {
            this.logger.info("at least one service is down");
            // at least one service down and we return failure
            resCode = responseCode.failure;
          }
          // Add self to the list
          globalResult.push({
            service: this.hcName,
            status: responseStatus.success,
            version: this.hcVersion
          });
          // Overall status of YLP depends on all services status
          ylp['status'] = (resCode == responseCode.success) ? responseStatus.success : responseStatus.failure;
          // Push YLP service information on first position
          globalResult.unshift(ylp);
          this.logger.debug(globalResult);
          // Respond
          res.status(resCode);
          res.send(globalResult);
        }
      });
    }
  }

  /**
   * Start the service
   * @param {*} port port on which the service should listen for incoming requests
   * @param {*} callback a function to call when the server started to listen
   */
  start(port, callback = null) {
    this.logger.debug("Starting health-check service");
    this.logger.debug("services array: " + JSON.stringify(this.services));
    //create server
    this.server = restify.createServer()
    //define routes
    this.server.get('/status', this.checkServices.bind(this));
    //and then start listening
    this.server.listen(port, () => {
      let serviceList = "";
      this.services.forEach(element => {
        serviceList += element.name + " : " + element.url + '\n';
      });
      this.logger.info("Health check service started on %s and will monitor the following services: \n %s", this.server.url, serviceList);
      if (callback != null) callback();
    });
  }

  /**
   * Stop the service
   */
  stop() {
    this.logger.debug("Stopping health-check service");
    this.server.close(() => {
      this.logger.info("Health Check service has stopped");
    })
  }
}

module.exports = HealthCheckService;
module.exports.responseStatus = responseStatus;
module.exports.responseCode = responseCode;
